src/parser.cpp

```diff
@@ -133,6 +133,8 @@ options_metadata parser::load_environment()
 
 options_metadata parser::load_settings()
 {
+    const auto dependent_config_setter = boost::bind(std::mem_fun(
+        &parser::set_dependent_variables), this);
```

Comment:
This is cool, but it has the issue I described earlier. The values will be incorrect if a setting field changes, and those fields cannot be made constant. Even if we were to assume cost-ness, the values are not set correctly in the case where this parser is not used (which is common). Also, computation of dependent fields cannot be accomplished on settings construct. So the only way to properly exposed computed properties is to make them functions, either on the settings class or externally.

Reply:
> The values will be incorrect if a setting field changes

Are you assuming that the settings will be dynamically changed after initialization? What is the use case for that? Some hard forks after a certain height?

> the values are not set correctly in the case where this parser is not used (which is common)

That's true. For btc these values are correct by construction for mainnet, testnet and regtest which are the main use cases. If someone is manipulating these values I'd assume that he knows what he's doing.

> So the only way to properly exposed computed properties is to make them functions, either on the settings class or externally.

You mean like [this](https://github.com/libbitcoin/libbitcoin/pull/1004#discussion_r209850494)? :p

Reply:
Yes, *all of the properties* computed from other properties on construct are problematic, since the other properties are not necessarily set at that point and computations are then too late. That's not the same as properties set from constants. Implementing the computation in that manner requires the user to reimplement it externally. I missed this initially.

Reply:
> which are the main use cases

Working properly only for the main use cases isn't generally sufficient for a class library.

Config values aren't limited to predefined defaults, they *do* change after initialization. There are existing cases where the defaults are overridden (see node and server). While it's less likely interesting to do so with consensus properties, the pattern should be the same.

The settings classes are not const, which makes fields computed only on initialization unsafe. This is why we changed all chain object fields to property methods, so we could recompute computed cached values (like hash) safely. We can't predict how people will use the classes. But we do ensure that read after write (and concurrent read - relating to cache) is correct for computed methods.